Pidgin.Examples/Expr/ExprParser.cs

```diff
@@ -0,0 +1,182 @@
+
+        public static readonly Parser<char, char> IdStart = UScore.Or(Letter);
+        public static readonly Parser<char, char> IdRest = UScore.Or(LetterOrDigit);
+        private static readonly Parser<char, string> IdInternal = Map((h, t) => h + t, IdStart, IdRest.ManyString());
+
+        public static readonly Parser<char, string> True = String("true").Labelled("true");
+        public static readonly Parser<char, string> False = String("false").Labelled("false");
+        public static readonly Parser<char, string> If = String("if").Labelled("if");
+        public static readonly Parser<char, string> Else = String("else").Labelled("else");
+        public static readonly Parser<char, string> Then = String("then").Labelled("then");
+        public static readonly Parser<char, string> In = String("in").Labelled("in");
+
+        private static readonly Parser<char, Unit> LookaheadKeyWordEnd = Lookahead(Not(IdRest).Or(End()));
+
+        public static readonly Parser<char, string> Bool = True.Or(False).Between(DQuote).Or(True.Or(False)).Labelled("bool");
+
+        public static readonly Parser<char, string> KeyWords =
+            OneOf(Else, Try(Then), Try(If), Try(In)).Before(LookaheadKeyWordEnd).Labelled("keyword");
+
+        public static readonly Parser<char, string> ReservedWords =
+            OneOf(False, Try(True), Else, Try(Then), Try(If), Try(In)).Before(LookaheadKeyWordEnd).Labelled("reserved");
+
+        public static readonly Parser<char, string> Id = Try(Lookahead(Not(ReservedWords).Or(End()))).Then(IdInternal).Labelled("id");
+
+        public static readonly Tuple<Parser<char, Expr>, Stack<bool>, Stack<bool>> Parser = BuildExprParser();
+
+        public static Result<char, Expr> Parse(string input)
+        {
+            try
+            {
+                var ret = Parser.Item1.Parse(input.Trim());
+                if (Parser.Item2.Count == 0 && Parser.Item3.Count == 0) return ret;
+
+                if (Parser.Item2.Count>0)
+                    throw new InvalidOperationException($"Unbalanced C-style Ternary Operators, still have {Parser.Item2.Count} unmatched.");
+                else
+                    throw new InvalidOperationException($"Unbalanced Python-style Ternary Operators, still have {Parser.Item3.Count} unmatched.");
+            }
+            finally
+            {
+                Parser.Item2.Clear();
+                Parser.Item3.Clear();
+            }
+        }
+
+        private static Tuple<Parser<char, Expr>, Stack<bool>, Stack<bool>> BuildExprParser()
+        {
+            Parser<char, Expr> exprParser = null;
+
+            var listParser = Rec(() => exprParser).Separated(Comma.Before(SkipWhitespaces));
+            var funcHead = Id.Or(If).Between(SkipWhitespaces).Before(LParen.Between(SkipWhitespaces));
+            var funcParser = Try(Lookahead(funcHead)).Then(Map((id, prms) => (Expr) new Function(id, prms), funcHead,
+                listParser.Before(RParen))).Labelled("func");
```

Comment:
I'd use a postfix operator for function invocations. Something like:

```csharp
// '(', then comma-separated expressions, then ')'
Operator.Postfix(
    Rec(() => expr)
        .SeparatedAtLeastOnce(Char(','))
        .Between(Char('('), Char(')'))
)
```

Reply:
The name of the function has to be an Id (or 'if' as a special case I supplied), it can't be a general expression.  Is Postfix operator workable?

Reply:
That’s not how it works in C#. `foo(bar)(baz)` or `((x, y) => x + y)(3,5)` are valid call expressions, for example

Reply:
You are right, I have been blinkered by my own specific use case of providing an expression language to users who are Excel power users (I don't expect them to use anonymous methods/lambdas).  Eventually, I still want to know the name of functions as distinct from other variables used in expressions as the users will use  growing library of functions we write in their expressions (in an Excel Addin).

Let me see what I can do with this.